syllabus/4-Smart_Contracts/6-Wasm_ink!_slides.md

```diff
@@ -1054,235 +1149,1255 @@ Notes:
+
+let receiver_balance_after = token.get_balance(receiver_one);
+
+assert_eq!(128, receiver_balance_after);
+```
+
+Notes:
+- money out of thin air
+
+---
+
+
+## The BatchOverflow exploit: fixit
+
+```rust [1|3-14]
+let amount = Self::safe_multiply(count as MyBalance, value)?;
+
+fn safe_multiply(a: MyBalance, b: MyBalance) -> Result<MyBalance, MyError> {
+    if a == 0 || b == 0 {
+        return Ok(0);
+    }
+
+    let c = a * b;
+    if c / a == b {
+        return Ok(c);
+    }
+
+    Err(MyError::Arithmetic)
+}
+
+```
+
+* We perform a reverse operation (division)
+* Value *c* is divided by *a* and we check whether this is equal to the value of *b*
+* If this was an overflow and wrapped around back to 0 then this check fails incorrect.
+
+---
+
+## The BatchOverflow exploit: fixit
+
+* runtime checks in Rust are performed only in *debug* mode, as they hinder performance
+```toml
+[profile.release]
+overflow-checks = false
+```
+
+* This prints 0 in *release* mode and panics in *debug*:
+```rust
+let x: u8 = "128".parse().unwrap();
+let val: u8 = x + 128;
+println!("{}", val);
+```
+
+* it's best to be explicit:
+```rust
+let amount = (count as MyBalance).checked_mul(value)
+              .ok_or(MyError::Arithmetic)?;
+```
+
+Notes:
+* in Rust there is no need to roll your own code
+* in Solidity take a look at the SafeMath library
+* explicit to the compiler about what you want to do
+
+---
+
+# MAXIMAL EXTRACTABLE VALUE (MEV)
+
+Notes:
+
+* we will now talk about MEV (Maximal extractable value)
+
+---
+
+# MEV
+
+* The concept of MEV was first floated as early as 2014 (Ethereum pre-genesis), in the context of Proof-of-work.
+* It was referred to as the *invisible tax*, the maximum value a miner can extract from moving around transactions when producing a block on a blockchain network.
+* After the Merge (Ethereum's move to POS consensus) *Miner Extractable Value* became *Maximal extractable value*.
+
+Notes:
+* MEV was first applied in the context of proof-of-work, and referred to as the Miner Extractable Value.
+* This is because in POW miners hold most of the power, controlling the transaction inclusion / exclusion and ordering.
+* However in the proof-of-stake the validators are been responsible for these roles.
+* The value extraction methods still exist though, so the term "Maximal extractable value" is now used instead.
+
+---
+
+## MEV: The Dark Forest
+
+* We have now seen a number of smart contract exploits.
+* Blockchain is a higly **competetive and adversarial** environment.
+* But the dangers pale in comparison to the **mempool**.
+
+Notes:
+* if a smart contract can be exploited for profit it's just a matter of time when it will be.
+* there is a lot af smart ppl spending a lot of time examining contracts for vulnerabilities.
+* mempool is the set of pending, unconfirmed transactions.
+
+---
+
+## MEV: The Dark Forest
+
+<img style="margin-top: 10px;margin-bottom: 10px" width="400" src="./img/ink/dark_forest.jpg" />
```

Comment:
Prefer to remove width and add height and addjust accordingly, as the text below the picture falls out of the screen (or maybe place image on left and text on the right on different columns)

Reply:
done